### ISSUE TITLE
HTTP2: Incorrect handling of frames received on closed stream

### ISSUE THREAD

**Issue by geoffkizer** [dotnet/corefx#34461]
Per RFC section 5.1:

```
Similarly, an endpoint
      that receives any frames after receiving a frame with the
      END_STREAM flag set MUST treat that as a connection error
      (Section 5.4.1) of type STREAM_CLOSED, unless the frame is
      permitted as described below
```

The EmptyResponse_FrameReceived_ResetsStream test is incorrect and should be changed.

@rmkerr

**Comment by geoffkizer** [comment]
Similar issue in the following tests:

CompletedResponse_FrameReceived_ResetsStream
ResetResponseStream_FrameReceived_ResetsStream

**Comment by rmkerr** [comment]
The spec gives this contradictory advice in section 6.1:
>If a DATA frame is received whose stream is not in "open" or "half-closed (local)" state, the recipient MUST respond with a stream error (Section 5.4.2) of type STREAM_CLOSED.

I guess that could mean we're supposed to send both, but that seems odd. Given that the advice in section 5.1 is more strict, I think that you're probably right and we should be sending a connection error.

**Comment by geoffkizer** [comment]
Yeah, the spec seems to be rather inconsistent about how to handle stream states in general, unfortunately.

I discovered this issue by trying to run the existing HTTP2 tests against WinHttpHandler. My inclination here is to make our behavior match WinHTTP, unless we have a really good reason not to. That said, I'm not sure it matters much, really, since these sorts of things are error cases and shouldn't happen in real world scenarios.

**Comment by rmkerr** [comment]
Does WinHttp respond with a GoAway frame before closing the connection? I'm trying to correct the test behavior now.

**Comment by geoffkizer** [comment]
I'm looking at this again. Strangely, it seems that WinHttp simply ignores the invalid frame. So the test hangs waiting for a RST_STREAM frame that never arrives, and eventually WinHttp closes the connection due to inactivity.

The connection still seems to be usable, despite the invalid frame. If I just modify the test to skip the logic that expects the RST_STREAM, and fall through to the next part of the test where we try to use the connection again, then the test passes.

I think I was confused before because I saw the connection being closed, and assumed that WinHttp was closing it because of the invalid frame. But it's not.

Given this and the spec ambiguity, I'm not quite sure what to do here. Thoughts?

**Comment by geoffkizer** [comment]
WinHttp also seems to ignore the invalid frames sent in the following tests:

DataFrame_IdleStream_ConnectionError
HeadersFrame_IdleStream_ConnectionError

**Comment by geoffkizer** [comment]
It's probably worthwhile to see what CurlHandler does here.

**Comment by rmkerr** [comment]
I'm not actually sure what we should do. The spec seems super ambiguous here. In particular I have issues with the following statements.

(1) The guidance on how to handle frames after a RST_STREAM:
>An endpoint that receives any frame other than PRIORITY after receiving a RST_STREAM MUST treat that as a stream error (Section 5.4.2) of type STREAM_CLOSED.

(2) The similar info for END_STREAM:
>Similarly, an endpoint that receives any frames after receiving a frame with the END_STREAM flag set MUST treat that as a connection error (Section 5.4.1) of type STREAM_CLOSED, unless the frame is permitted as described below.

Both of the above (Receiving a RST_STREAM or an END_STREAM) cause the stream to go into the CLOSED state, but the spec expects us to respond in a different way if an additional frame is sent on the connection. Which seems to imply that we should be retaining some state info for each stream on how it was closed? That just seems like a bad idea to me.

It's clearly not in line with the spec or with WinHttp, but my first inclination would be to respond with a connection error (and potentially a GoAway frame).

I agree that we should do some more testing though.

**Comment by geoffkizer** [comment]
> Which seems to imply that we should be retaining some state info for each stream on how it was closed? That just seems like a bad idea to me.

Totally agree

> It's clearly not in line with the spec or with WinHttp, but my first inclination would be to respond with a connection error (and potentially a GoAway frame).

I'm inclined to agree. If a peer sends an invalid frame, it seems like it should be a connection error.

**Comment by davidsh** [comment]
>Given this and the spec ambiguity, I'm not quite sure what to do here. Thoughts?

@geoffkizer It would be useful to pass this feedback and questions to the WinHTTP team directly.  At least get their perspective on why it was implemented that way.  Send me email offline if you need a contact person for that.

**Comment by rmkerr** [comment]
I have a change ready that implements that behavior (along with GoAway). I'll open up a PR if anyone wants to take a look, but will leave it marked as WIP until we hear from the WinHttp team.

**Comment by geoffkizer** [comment]
I sent mail to the WinHTTP folks, you guys are copied.

I will also post a PR so we can see what happens with CurlHandler.

**Comment by karelz** [comment]
@rmkerr do we need anything else here? Is it must-have for 3.0?

**Comment by rmkerr** [comment]
Closing, as most of the issues here have been addressed in #34629, and the others have been split out into #35648.
